### ISSUE TITLE
为什么我用Bert-pytorch 生成的词向量 与 bert-as-service生成的词向量不同呢，都是用的中文

### ISSUE THREAD

**Issue by duan348733684** [hanxiao/bert-as-service#498]
为什么我用Bert-pytorch 生成的词向量 与 bert-as-service生成的词向量不同呢？都是用的中文。。。感谢回答

**Comment by hzk123** [comment]
pytorch要用train.eval()把dropout关掉

**Comment by duan348733684** [comment]
> pytorch要用train.eval()把dropout关掉

################################################
import warnings
warnings.filterwarnings("ignore")
import torch
from transformers import BertTokenizer, BertModel

UNCASE = "./bert-base-chinese"
VOCAB = "vocab.txt"
tokenizer = BertTokenizer.from_pretrained(UNCASE + "/" + VOCAB)
model = BertModel.from_pretrained(UNCASE + "/test")
model.eval()
string = '盘踞在山中的妖魔'
# string = "[CLS]" + string + "[SEP]"
tokenized_string = tokenizer.tokenize(string)
tokens_ids = tokenizer.convert_tokens_to_ids(tokenized_string)
tokens_tensor = torch.tensor([tokens_ids])
outputs = model(tokens_tensor)
print(outputs[1])

###############################################
import warnings
warnings.filterwarnings('ignore')
from bert_serving.client import BertClient
bc_ch = BertClient(ip='algo.haetek.com', port=5555, port_out=5556)  # chinese
vec = bc_ch.encode(['盘踞在山中的妖魔'])
print(vec)

################################################
上面分别是pytorch-bert和bert-as-service的生成词向量的方式，，最后出来的词向量不同，我加了eval()啊。请您指教一下哪里出问题了

**Comment by hksskh** [comment]
bert-as-service有几个参数可能影响结果：

**Comment by hksskh** [comment]
bert-as-service有几个参数可能影响结果：
-pooling_layer，默认是[-2]，取的是倒数第二层的表示
-pooling_strategy，默认是REDUCE_MEAN
你核对一下这些配置，和你pytorch模型的配置是否一致

**Comment by duan348733684** [comment]
> bert-as-service有几个参数可能影响结果：
> -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> -pooling_strategy，默认是REDUCE_MEAN
> 你核对一下这些配置，和你pytorch模型的配置是否一致

谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？

**Comment by hksskh** [comment]
> > bert-as-service有几个参数可能影响结果：
> > -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> > -pooling_strategy，默认是REDUCE_MEAN
> > 你核对一下这些配置，和你pytorch模型的配置是否一致
> 
> 谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？

pooling_layer为-1表示最后用来预测MLM和NSP的那一层输出
bert一共有12个transformer块，pooling_layer为-2表示取的是第11个块的输出，你可以看一下bert源码modeling.py里的all_layer_outputs这个列表的内容。

又发现了几个影响输出embedding的参数：
max_seq_len（太短有可能把你的输入截断了）
mask_cls_sep（可以选择把[CLS]和[SEP]给mask掉，不参与reduce_mean）

另外，看你贴出来的代码，如果按默认配置，bert-as-service出来的是倒数第二个transformer块的词向量平均；但是我不太清楚pytorch那部分的outputs[1]输出的是啥。。。

**Comment by liushaoweihua** [comment]
作者的做法是把输出整句做了reduce_mean，包含[[CLS],[],...,[SEP],[],...,[SEP]]，另一个取的是[CLS]的向量。

**Comment by duan348733684** [comment]
> > > bert-as-service有几个参数可能影响结果：
> > > -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> > > -pooling_strategy，默认是REDUCE_MEAN
> > > 你核对一下这些配置，和你pytorch模型的配置是否一致
> > 
> > 
> > 谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？
> 
> pooling_layer为-1表示最后用来预测MLM和NSP的那一层输出
> bert一共有12个transformer块，pooling_layer为-2表示取的是第11个块的输出，你可以看一下bert源码modeling.py里的all_layer_outputs这个列表的内容。
> 
> 又发现了几个影响输出embedding的参数：
> max_seq_len（太短有可能把你的输入截断了）
> mask_cls_sep（可以选择把[CLS]和[SEP]给mask掉，不参与reduce_mean）
> 
> 另外，看你贴出来的代码，如果按默认配置，bert-as-service出来的是倒数第二个transformer块的词向量平均；但是我不太清楚pytorch那部分的outputs[1]输出的是啥。。。

谢谢指导，我得到了相同的向量。。。再请教一下，这里这个-max_seq_len 的默认值是 25， 如果sequence小于这个默认值的话，会padding到长度为25吗? padding 0 还是 1?

**Comment by hksskh** [comment]
> > > > bert-as-service有几个参数可能影响结果：
> > > > -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> > > > -pooling_strategy，默认是REDUCE_MEAN
> > > > 你核对一下这些配置，和你pytorch模型的配置是否一致
> > > 
> > > 
> > > 谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？
> > 
> > 
> > pooling_layer为-1表示最后用来预测MLM和NSP的那一层输出
> > bert一共有12个transformer块，pooling_layer为-2表示取的是第11个块的输出，你可以看一下bert源码modeling.py里的all_layer_outputs这个列表的内容。
> > 又发现了几个影响输出embedding的参数：
> > max_seq_len（太短有可能把你的输入截断了）
> > mask_cls_sep（可以选择把[CLS]和[SEP]给mask掉，不参与reduce_mean）
> > 另外，看你贴出来的代码，如果按默认配置，bert-as-service出来的是倒数第二个transformer块的词向量平均；但是我不太清楚pytorch那部分的outputs[1]输出的是啥。。。
> 
> 谢谢指导，我得到了相同的向量。。。再请教一下，这里这个-max_seq_len 的默认值是 25， 如果sequence小于这个默认值的话，会padding到长度为25吗? padding 0 还是 1?

padding添加0；
当句子长度小于25时，padding的策略分两种：
有这个参数：
`group2.add_argument('-no_special_token', action='store_true', default=False,
                        help='add [CLS] and [SEP] in every sequence, \
                        put sequence to the model without [CLS] and [SEP] when True and \
                        is_tokenized=True in Client')`
这个参数表示是不是在网络中使用[CLS]和[SEP]；如果不用，那么直接把你的句子padding到长度25；如果用，先在句子头尾加这两个标志，如果长度还不够，再padding到25

详见server/bert_serving/server/bert下的extract_features中的convert_lst_to_features函数

**Comment by duan348733684** [comment]
> > > > > bert-as-service有几个参数可能影响结果：
> > > > > -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> > > > > -pooling_strategy，默认是REDUCE_MEAN
> > > > > 你核对一下这些配置，和你pytorch模型的配置是否一致
> > > > 
> > > > 
> > > > 谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？
> > > 
> > > 
> > > pooling_layer为-1表示最后用来预测MLM和NSP的那一层输出
> > > bert一共有12个transformer块，pooling_layer为-2表示取的是第11个块的输出，你可以看一下bert源码modeling.py里的all_layer_outputs这个列表的内容。
> > > 又发现了几个影响输出embedding的参数：
> > > max_seq_len（太短有可能把你的输入截断了）
> > > mask_cls_sep（可以选择把[CLS]和[SEP]给mask掉，不参与reduce_mean）
> > > 另外，看你贴出来的代码，如果按默认配置，bert-as-service出来的是倒数第二个transformer块的词向量平均；但是我不太清楚pytorch那部分的outputs[1]输出的是啥。。。
> > 
> > 
> > 谢谢指导，我得到了相同的向量。。。再请教一下，这里这个-max_seq_len 的默认值是 25， 如果sequence小于这个默认值的话，会padding到长度为25吗? padding 0 还是 1?
> 
> padding添加0；
> 当句子长度小于25时，padding的策略分两种：
> 有这个参数：
> `group2.add_argument('-no_special_token', action='store_true', default=False, help='add [CLS] and [SEP] in every sequence, \ put sequence to the model without [CLS] and [SEP] when True and \ is_tokenized=True in Client')`
> 这个参数表示是不是在网络中使用[CLS]和[SEP]；如果不用，那么直接把你的句子padding到长度25；如果用，先在句子头尾加这两个标志，如果长度还不够，再padding到25
> 
> 详见server/bert_serving/server/bert下的extract_features中的convert_lst_to_features函数

谢谢您，，，你有用transformers里的Bert吗？

**Comment by duan348733684** [comment]
> > > > > bert-as-service有几个参数可能影响结果：
> > > > > -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> > > > > -pooling_strategy，默认是REDUCE_MEAN
> > > > > 你核对一下这些配置，和你pytorch模型的配置是否一致
> > > > 
> > > > 
> > > > 谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？
> > > 
> > > 
> > > pooling_layer为-1表示最后用来预测MLM和NSP的那一层输出
> > > bert一共有12个transformer块，pooling_layer为-2表示取的是第11个块的输出，你可以看一下bert源码modeling.py里的all_layer_outputs这个列表的内容。
> > > 又发现了几个影响输出embedding的参数：
> > > max_seq_len（太短有可能把你的输入截断了）
> > > mask_cls_sep（可以选择把[CLS]和[SEP]给mask掉，不参与reduce_mean）
> > > 另外，看你贴出来的代码，如果按默认配置，bert-as-service出来的是倒数第二个transformer块的词向量平均；但是我不太清楚pytorch那部分的outputs[1]输出的是啥。。。
> > 
> > 
> > 谢谢指导，我得到了相同的向量。。。再请教一下，这里这个-max_seq_len 的默认值是 25， 如果sequence小于这个默认值的话，会padding到长度为25吗? padding 0 还是 1?
> 
> padding添加0；
> 当句子长度小于25时，padding的策略分两种：
> 有这个参数：
> `group2.add_argument('-no_special_token', action='store_true', default=False, help='add [CLS] and [SEP] in every sequence, \ put sequence to the model without [CLS] and [SEP] when True and \ is_tokenized=True in Client')`
> 这个参数表示是不是在网络中使用[CLS]和[SEP]；如果不用，那么直接把你的句子padding到长度25；如果用，先在句子头尾加这两个标志，如果长度还不够，再padding到25
> 
> 详见server/bert_serving/server/bert下的extract_features中的convert_lst_to_features函数

谢谢您，，，您有用transformers里的bert吗  用torch实现的

**Comment by hksskh** [comment]
> > > > > > bert-as-service有几个参数可能影响结果：
> > > > > > -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> > > > > > -pooling_strategy，默认是REDUCE_MEAN
> > > > > > 你核对一下这些配置，和你pytorch模型的配置是否一致
> > > > > 
> > > > > 
> > > > > 谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？
> > > > 
> > > > 
> > > > pooling_layer为-1表示最后用来预测MLM和NSP的那一层输出
> > > > bert一共有12个transformer块，pooling_layer为-2表示取的是第11个块的输出，你可以看一下bert源码modeling.py里的all_layer_outputs这个列表的内容。
> > > > 又发现了几个影响输出embedding的参数：
> > > > max_seq_len（太短有可能把你的输入截断了）
> > > > mask_cls_sep（可以选择把[CLS]和[SEP]给mask掉，不参与reduce_mean）
> > > > 另外，看你贴出来的代码，如果按默认配置，bert-as-service出来的是倒数第二个transformer块的词向量平均；但是我不太清楚pytorch那部分的outputs[1]输出的是啥。。。
> > > 
> > > 
> > > 谢谢指导，我得到了相同的向量。。。再请教一下，这里这个-max_seq_len 的默认值是 25， 如果sequence小于这个默认值的话，会padding到长度为25吗? padding 0 还是 1?
> > 
> > 
> > padding添加0；
> > 当句子长度小于25时，padding的策略分两种：
> > 有这个参数：
> > `group2.add_argument('-no_special_token', action='store_true', default=False, help='add [CLS] and [SEP] in every sequence, \ put sequence to the model without [CLS] and [SEP] when True and \ is_tokenized=True in Client')`
> > 这个参数表示是不是在网络中使用[CLS]和[SEP]；如果不用，那么直接把你的句子padding到长度25；如果用，先在句子头尾加这两个标志，如果长度还不够，再padding到25
> > 详见server/bert_serving/server/bert下的extract_features中的convert_lst_to_features函数
> 
> 谢谢您，，，您有用transformers里的bert吗 用torch实现的

没用过torch。。

**Comment by MrRace** [comment]
> > > > bert-as-service有几个参数可能影响结果：
> > > > -pooling_layer，默认是[-2]，取的是倒数第二层的表示
> > > > -pooling_strategy，默认是REDUCE_MEAN
> > > > 你核对一下这些配置，和你pytorch模型的配置是否一致
> > > 
> > > 
> > > 谢谢回复，这个倒数第二层 是指transformer 的倒数第二层吗？
> > 
> > 
> > pooling_layer为-1表示最后用来预测MLM和NSP的那一层输出
> > bert一共有12个transformer块，pooling_layer为-2表示取的是第11个块的输出，你可以看一下bert源码modeling.py里的all_layer_outputs这个列表的内容。
> > 又发现了几个影响输出embedding的参数：
> > max_seq_len（太短有可能把你的输入截断了）
> > mask_cls_sep（可以选择把[CLS]和[SEP]给mask掉，不参与reduce_mean）
> > 另外，看你贴出来的代码，如果按默认配置，bert-as-service出来的是倒数第二个transformer块的词向量平均；但是我不太清楚pytorch那部分的outputs[1]输出的是啥。。。
> 
> 谢谢指导，我得到了相同的向量。。。再请教一下，这里这个-max_seq_len 的默认值是 25， 如果sequence小于这个默认值的话，会padding到长度为25吗? padding 0 还是 1?

方便贴下可以pytorch-bert获得与bert-as-service一样向量的完整代码吗？同样在实验。谢谢

**Comment by zhaoguangyao** [comment]
@duan348733684 请问控制哪些参数使得pytorch的输出结果与bert as service一致？谢谢。

**Comment by MrRace** [comment]
> @duan348733684 请问控制哪些参数使得pytorch的输出结果与bert as service一致？谢谢。

https://blog.csdn.net/ljp1919/article/details/106592828

**Comment by MrRace** [comment]
@zhaoguangyao  https://blog.csdn.net/ljp1919/article/details/106592828

**Comment by zhaoguangyao** [comment]
> @zhaoguangyao https://blog.csdn.net/ljp1919/article/details/106592828

thanks.

**Comment by Apeoud** [comment]
> @zhaoguangyao https://blog.csdn.net/ljp1919/article/details/106592828

感谢分享！不过需要注意的是，如果padding的话，不能直接torch.mean，需要感觉mask来mean，不然还是和bert as service的不一致。
